### PR TITLE
feat: have `landscape-config --silent` only send a registration messa…

### PR DESCRIPTION
**Pull request by wck0**
…ge when required and add a new `--force-registration` flag

### CODE REVIEW

**Comment by wck0** [comment]
Currently, `landscape-config --silent` will always try to send a registration message. With this change, when the `--silent` flag is set we only send the registration message if we can deduce that no registration message has been sent yet (i.e. the secure id is null), or the `landscape-config` is explicitly called with the `--force-registration` flag.

Manual testing instructions:

1. Working on this branch, configure Landscape Client and have it send a registration message to create a new pending computer.
2. Run `landscape-config --silent` and observe no additional pending computer is created.
3. Accept the pending computer and allow some message exchange to occur.
4. Run `landscape-config --silent` and observe no additional pending computer is created.
5. Run `landscape-config --silent --force-registration` and observe no additional pending computer is created.

Repeat ad nauseam making changes to the configuration along the way.

**Comment by srunde3** [comment]
Why do we want `--force-registration` to *not* create a new pending computer?  Wouldn't we want it to create a new secure_id for the client and therefore a new pending computer?

**Review by srunde3 — COMMENTED** [review]
Manual testing went fine according to your instructions; left one question about if your documented behavior is what we "want" from this change

**Comment by silverdrake11** [comment]
How is this different from this pull request https://github.com/canonical/landscape-client/pull/243

**Comment by silverdrake11** [comment]
One other question. Will this break existing scripts that rely on landscape-config sending a new registration as the default behavior? I can see that the charm will be affected by this. For example, if the user runs the re-register run action, then that is implement simply by a `landscape-config --silent`

**Comment by wck0** [comment]
> Why do we want `--force-registration` to _not_ create a new pending computer? Wouldn't we want it to create a new secure_id for the client and therefore a new pending computer?

We don't. I made a copy-pasta error in my testing instructions. Editing them to better reflect the intentions.

**Comment by wck0** [comment]
> How is this different from this pull request #243

The `--skip-registration` flag can be passed independent of the `--silent` flag.

**Comment by wck0** [comment]
> One other question. Will this break existing scripts that rely on landscape-config sending a new registration as the default behavior? I can see that the charm will be affected by this. For example, if the user runs the re-register run action, then that is implement simply by a `landscape-config --silent`

Yes, it sounds like this change will break that behavior. It can be adjusted in the charm by calling `landscape-config --silent --force-registration` if the charm action should always result in a new registration request.

**Comment by silverdrake11** [comment]
In that case I would prefer we have a flag `--send-only-if-unsent` (maybe we can think of a better name) than modify the default behavior of landscape client. We have alot of customers relying on old default behavior and changing the api in a way that could result in breaking changes seems like it can cause trouble

**Comment by wck0** [comment]
OK, good point. I'll fix this up and let Carlos know.

**Comment by wck0** [comment]
Leaning towards `--register-if-needed` but open to suggestions.

**Comment by wck0** [comment]
Updated, restoring the behavior of `--silent` and with a new flag `--register-if-needed`.
Other refactoring, too.

File: landscape/client/configuration.py [hunk 1]
```diff
@@ -922,9 +948,17 @@ def main(args, print=print):
...
+        sys.exit(
+            "Do not set both skip registration "
+            "and force registration together.",
+        )
+
+    already_registered = is_registered(config)
+
     if config.is_registered:
 
-        registration_status = is_registered(config)
```
[the remark this answers is not in the document]
Comment (wck0) [reply landscape/client/configuration.py:927]: I think I picked this nit.

**Review by srunde3 — COMMENTED** [review]
Code looks good, but I'm getting unexpected registration messages after running `--silent --register-if-needed`.  I'll continue this tomorrow to see if the issue exists between chair and keyboard or if changes are needed here.

**Review by srunde3 — CHANGES_REQUESTED** [review]

**Comment by wck0** [comment]
Updated. For manual testing now:

1.  start with a fresh snap install
2. register with your Landscape Server
3. accept the pending computer
4. wait for message exchange to occur. This ensures that the `secure_id` is set on the client to match the accepted computer
5. inspect the `broker.bpickle` file and note the value of `secure_id`
6. run `landscape-config --silent --register-if-needed`
7. observe no new pending computer is created.
8. inspect the `broker.bpickle` file and note that the value of `secure_id` has not changed
9. try to break something

File: landscape/client/configuration.py [hunk 2]
```diff
@@ -953,31 +997,37 @@ def main(args, print=print):
...
-
     if config.skip_registration:
+        result = "registration-skipped"
+        report_registration_outcome(result, print=print)
         return
 
-    if config.silent:
+    should_register = False
+
+    if config.force_registration:
```
Comment (srunde3) [inline landscape/client/configuration.py:1007]: nit: you could rearrange this block to make it simpler, if you so desire:

```py
    if config.force_registration:
        should_register = True
    elif config.register_if_needed:
        should_register = not already_registered
    elif config.silent:
        should_register = True
    else:
        ...
```

**Review by srunde3 — APPROVED** [review]
LGTM! :ship: left one pedantic inline nit for your viewing pleasure.  Manual testing went well.

**Pull request merged**